pkg/health/condition/check_ready.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package condition
+
+import druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+
+type readyCheck struct{}
+
+func (r *readyCheck) Check(status druidv1alpha1.EtcdStatus) Result {
+	if len(status.Members) == 0 {
+		return &result{
+			conType: druidv1alpha1.ConditionTypeReady,
+			status:  druidv1alpha1.ConditionUnknown,
+			reason:  "NoMembersInStatus",
+			message: "Cannot determine readiness of cluster since status has no members",
+		}
+	}
+
+	var (
+		quorum       = len(status.Members)/2 + 1
```

Comment:
```suggestion
		quorum       = status.Replicas/2 + 1
```

Reply:
Good point, I deliberately chose `status.Member` because it reflects the current state of the cluster. Let's imagine you change `spec.replicas` from `1 -> 3`, but the two new pods will remain in `Pending` for some time.
1. Using `status.Members`, the condition will still be `Status: "True"` because it underlying etcd cluster still only exists of one member, i.e., quorum is given.
2. Using `status.Replicas` then it'll be `Status: "False"`.

The question is if we want to rather reflect the cluster state (1.) or refer to what's given in the `.spec` (2.).

What do you think?

Reply:
To summarise my understanding of quorum in an etcd cluster, `quorum = cluster_size/2 + 1` where `cluster_size` is the current number of members (excluding learners) in the cluster. While scaling `n -> m` where `n > 1 and m > 1` this is quite straightforward. So, using `len(status.members)` or `status.Replicas` may not make much of a difference provided scaling is always done on fully healthy clusters (`AllMembersReady`). There might be a difference if we scale even when all members are not ready (but the cluster is quorate).

But it gets a bit more complicated while bootstrapping an etcd cluster (empty or from backup) because then the `cluster_size` is determined by the number of peer URLs passed as `ETCD_INITIAL_CLUSTER`. The case of `1 -> n` is a special case because it is really a bootstrap scenario. In that case (as well as the usual cluster bootstrap case), I thought `status.Replicas` may reflect the contents of `ETCD_INITIAL_CLUSTER` better than `len(status.members)` because it will take time for the pods to come up and their sidecars to populate `status.members`. During that time the cluster is really not quorate.

In the case of scaling `n -> m` where `n > 1 and m > 1` when all members are not ready, using `status.Replicas` might be more accurate as well.

WDYT?

Reply:
> In the case of scaling n -> m where n > 1 and m > 1 when all members are not ready, using status.Replicas might be more accurate as well.

On second thought, `len(status.members)` might be a better measure of quorum in this case. So, in summary, the bootstrap case might require `status.Replicas` to be considered.

Reply:
No sure if we can detect the bootstrap case with current fields in the status. Might we need to introduce a `Bootstrapping` condition type?

Reply:
cc @shreyas-s-rao @abdasgupta @ishan16696

Reply:
I gave it some more thought. According to the [Kubernetes API reference](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/stateful-set-v1/#StatefulSetStatus)...

> replicas is the number of Pods created by the StatefulSet controller.

So, it indicates the number of pods that are present at any point in time. I think calculating quorum based on this is equally problematic during bootstrap (e.g., if initial cluster size is 5, we might wrongly declare that the cluster is quorate already when 2 replicas are ready).

Using `spec.Replicas` instead would lead to wrong quorum checks for the the case of `n -> m` where `n > 1 and m > 1` (e.g. when scaling `3  -> 7`, we might wrongly declare that the cluster is not quorate until there are 4 ready replicas). 

Adding a `Bootstrapping` condition type can help, but will make the logic complicated to both calculate the condition as well as to consider it in the `Ready` condition checker to use `spec.Replicas` for cluster size in that case.

Another alternative is to introduce a new field `status.initialReplicas` which can be set to `spec.Replicas` by `etcd-druid` whenever bootstrapping triggered. Then the `Ready` condition checker can consider `max(status.IntialReplicas, status.Replicas)` as the cluster size checking calculating quorum.

WDYT?

Reply:
> Using spec.Replicas instead would lead to wrong quorum checks for the the case of n -> m where n > 1 and m > 1 (e.g. when scaling 3 -> 7, we might wrongly declare that the cluster is not quorate until there are 4 ready replicas).

I think the same is true for `status.replicas` because as mentioned in the example of my previous comment, the count might be updated quite fast (i.e. Pod created) but then depending on different factors, it might take longer for the pod to be started (e.g. long disk attachment time) and the member to eventually to join the cluster the first time. I don't think that we want to report a quorum loss in the meantime because (a) there is still a quorum, (b) potential remediation (e.g. killing the pod) can end in a worse situation.

From what I can see:
- Druid does know very little about a running etcd cluster
- It should rely on information provided by the etcd-br side car through `etcd.status.members[]`
- Thus, the etcd-br side car should update the corresponding member as soon as possible
- For bootstrapping I either like your `status.intialReplicas` or we follow the [status.lastOperation](https://github.com/gardener/gardener/blob/0968f39752e046235ed1bdf4618781b6b57268c9/pkg/apis/core/v1beta1/types_common.go#L101) approach which we already use for the most of our Gardener resources. The `status.lastOperation.type` will identify if the cluster is bootstrapping (`Create`/`Restore`) or not.

Reply:
> I think the same is true for status.replicas

Agreed about the problem with using `status.Replicas`. We should use `len(status.Members)` like you have done in the PR 👍 

> For bootstrapping I either like your status.intialReplicas or we follow the status.lastOperation approach which we already use for the most of our Gardener resources. The status.lastOperation.type will identify if the cluster is bootstrapping (Create/Restore) or not.

I am thinking that we might need both. Only using `status.lastOperation` would mean that we have to use `spec.Replicas` as initial replicas in the bootstrapping case. But this might be incorrect, if `spec.replicas` is updated while bootstrapping is on-going. So, it might be safer for druid to copy `spec.replicas` to `status.initialReplicas` at the beginning of triggering bootstrapping. Though strictly not necessary `status.lastOperation` might be more useful for human users to understand what is going on.

Reply:
> But this might be incorrect, if spec.replicas is updated while bootstrapping is on-going

You have point there which at the same time brings me to my next question: Should we even exit the reconciliation loop as long as not all pods are ready after a spec change (e.g. `spec.replicas: 1` -> `spec.replicas: 3`)? I like the short cycle that the Druid has today but OTHO waiting until everything is up-and-running also has its benefits, e.g., a `Processing` operation signals that the cluster is still "under construction" and thus conditions can flap in the meantime.

Reply:
> waiting until everything is up-and-running also has its benefits, e.g., a Processing operation signals that the cluster is still "under construction" and thus conditions can flap in the meantime.

I find long reconciliations against the _spirit_ of control-loops, especially, if they are reliant on such long reconciliation loops to not be interrupted. E.g. while waiting for the cluster to bootstrap the druid pod might restart and then it will need a way to compute the current state during the next reconciliation. Using `spec.replicas` at that time could be just as wrong. So, something like `status.intialReplicas` will be still required (but only in rarer cases). I think solutions that work for short reconciliation loops can be simpler because of lack of in-memory state and work better in general.

But this is in no way an argument against `Processing` or `Bootstrapping` operation type in `status.lastOperation`. In fact, it is an argument that we should have it but along with `status.initialReplicas` and short reconcilations.

Reply:
I have updated the solution in the proposal in #189 based on our out-of-band discussion.